plugins/in_tcp/tcp_conn.c

```diff
@@ -50,22 +119,72 @@ static inline int process_pack(struct tcp_conn *conn,
     while (msgpack_unpack_next(&result, pack, size, &off) == MSGPACK_UNPACK_SUCCESS) {
         entry = result.data;
 
+        appended_address_buffer = NULL;
+        source_address = NULL;
+
         ret = flb_log_event_encoder_begin_record(ctx->log_encoder);
 
         if (ret == FLB_EVENT_ENCODER_SUCCESS) {
             ret = flb_log_event_encoder_set_current_timestamp(ctx->log_encoder);
         }
 
+        if (ctx->source_address_key != NULL) {
+            source_address = flb_connection_get_remote_address(conn->connection);
+        }
+
         if (ret == FLB_EVENT_ENCODER_SUCCESS) {
             if (entry.type == MSGPACK_OBJECT_MAP) {
-                ret = flb_log_event_encoder_set_body_from_msgpack_object(
-                        ctx->log_encoder, &entry);
+                if (source_address != NULL) {
+                    msgpack_sbuffer_init(&sbuf);
+                    msgpack_packer_init(&pck, &sbuf, msgpack_sbuffer_write);
+
+                    len = entry.via.map.size;
+                    msgpack_pack_map(&pck, len);
+
+                    for (i=0; i<len; i++) {
+                        msgpack_pack_object(&pck, entry.via.map.ptr[i].key);
+                        msgpack_pack_object(&pck, entry.via.map.ptr[i].val);
+                    }
+
+                    ret = append_message_to_record_data(&appended_address_buffer,
+                                                        &appended_address_size,
+                                                        ctx->source_address_key,
+                                                        sbuf.data,
+                                                        sbuf.size,
+                                                        source_address,
+                                                        strlen(source_address),
+                                                        MSGPACK_OBJECT_STR);
+                    msgpack_sbuffer_destroy(&sbuf);
+                }
+
+                if (ret == FLB_MAP_EXPANSION_ERROR) {
+                    flb_plg_debug(ctx->ins, "error expanding source_address : %d", ret);
+                }
+
+                if (appended_address_buffer != NULL) {
+                    ret = flb_log_event_encoder_set_body_from_raw_msgpack(
+                            ctx->log_encoder, appended_address_buffer, appended_address_size);
+                }
+                else {
+                    ret = flb_log_event_encoder_set_body_from_msgpack_object(
+                            ctx->log_encoder, &entry);
+                }
             }
             else if (entry.type == MSGPACK_OBJECT_ARRAY) {
-                ret = flb_log_event_encoder_append_body_values(
+                if (source_address != NULL) {
```

Comment:
Can't `ctx->source_address_key` be `NULL`? I think it could and we should handle it in that case.